pxr/usd/usd/wrapValidator.cpp

```diff
@@ -0,0 +1,106 @@
+//
+// Licensed under the terms set forth in the LICENSE.txt file available at
+// https://openusd.org/license.
+//
+#include "pxr/pxr.h"
+#include "pxr/usd/usd/validator.h"
+
+#include "pxr/base/tf/pyContainerConversions.h"
+#include "pxr/base/tf/pyPtrHelpers.h"
+#include "pxr/base/tf/pyResultConversions.h"
+
+#include <boost/python/class.hpp>
+#include <boost/python/make_constructor.hpp>
+#include <boost/python/operators.hpp>
+#include <boost/python/object.hpp>
+
+using namespace boost::python;
+
+PXR_NAMESPACE_USING_DIRECTIVE
+
+namespace
+{
+
+    UsdValidatorMetadata *
+    _NewMetadata(
+        const TfToken &name,
+        const PlugPluginPtr &plugin,
+        const TfTokenVector &keywords,
+        const TfToken &doc,
+        const TfTokenVector &schemaTypes,
+        bool isSuite)
+    {
+        return new UsdValidatorMetadata{name, plugin, keywords, doc, schemaTypes, isSuite};
+    }
+
+    TfToken
+    _GetMetadataName(const UsdValidatorMetadata &metadata)
+    {
+        return metadata.name;
+    }
+
+    void
+    _SetMetadataName(UsdValidatorMetadata &metadata, const TfToken &name)
+    {
+        metadata.name = name;
+    }
+
+    TfTokenVector
+    _GetMetadataKeywords(const UsdValidatorMetadata &metadata)
+    {
+        return metadata.keywords;
+    }
+
+    void
+    _SetMetadataKeywords(UsdValidatorMetadata &metadata, const TfTokenVector &keywords)
+    {
+        metadata.keywords = keywords;
+    }
+
+    TfTokenVector
+    _GetMetadataSchemaTypes(const UsdValidatorMetadata &metadata)
+    {
+        return metadata.schemaTypes;
+    }
+
+    void
+    _SetMetadataSchemaTypes(UsdValidatorMetadata &metadata, const TfTokenVector &schemaTypes)
+    {
+        metadata.schemaTypes = schemaTypes;
+    }
+
+    PlugPluginPtr
+    _GetMetadataPlugin(const UsdValidatorMetadata &metadata)
+    {
+        return metadata.pluginPtr;
+    }
+
+    void
+    _SetMetadataPlugin(UsdValidatorMetadata &metadata, const PlugPluginPtr &plugin)
+    {
+        metadata.pluginPtr = plugin;
+    }
+
+} // anonymous namespace
+
+void wrapUsdValidator()
+{
+    class_<UsdValidatorMetadata>("ValidatorMetadata", no_init)
+        .def("__init__", make_constructor(&_NewMetadata, default_call_policies(),
+                                          (arg("name") = TfToken(),
+                                           arg("plugin") = PlugPluginPtr(),
+                                           arg("keywords") = TfTokenVector(),
+                                           arg("doc") = TfToken(),
+                                           arg("schemaTypes") = TfTokenVector(),
+                                           arg("isSuite") = false)))
+        .add_property("name", &_GetMetadataName, &_SetMetadataName)
+        .add_property("plugin", &_GetMetadataPlugin, &_SetMetadataPlugin)
+        .add_property("keywords", make_function(
+            &_GetMetadataKeywords, return_value_policy<TfPySequenceToList>()), _SetMetadataKeywords)
```

Comment:
I was having some good discussion with Alex here at Pixar about python APIs here and how it might be confusing for a python developer to not be able to do something like: `metadata.keywords[0] = "foo"`. We also discussed and have many examples of this in our code base where we wrap something as a python property if value of the type is immutable in python.

So that gets me to, will it make more sense to provide a GetKeywords/SetKeywords (and similarly for schemaTypes below) which returns a list and takes in a list respectively? (We have example of this here: https://github.com/PixarAnimationStudios/OpenUSD/blob/release/pxr/usd/usd/wrapAttribute.cpp#L123-L124

But now this results in inconsistency with the cpp API, which at present doesn't provide any member functions on the UsdValidationMetadata struct. Do you think it will be possible for you to add these methods also for consistency?

Thanks

Reply:
That makes sense to me. I'll change it to Set/Get function instead. Thanks.

Reply:
@tallytalwar I have changed it to use getter/setter to access keywords and schemaTypes, but not for cpp struct. One question is that should I keep `keywords` and `schemaTypes` as private or add public getter/setter for them only? That adds another inconsistency that it has two set of methods to access the two members if I keep them public. While it breaks back compatibility if I moved them into private.

Reply:
What about using a `vector_indexing_suite` to preserve consistency with the CPP API?

Reply:
Rather than use `vector_indexing_suite` or adding explicit get/set methods, I think I'd actually argue in favor of just making `UsdVaidatorMetadata` immutable in python.  In general, the usage for `UsdValidatorMetadata` is to just reflect and inspect metadata described in C++ validator plugins.

There's already a keyword argument interface for easy construction for testing.  `keywords` and `schemaTypes` could use `TfPySequenceToTuple`.

Reply:
Yes, that makes sense to me if we don't need to register validator in python.

Reply:
Hmm, I think we should allow registering of validators from python also. So I can see we providing API for this https://github.com/PixarAnimationStudios/OpenUSD/blob/release/pxr/usd/usd/validationRegistry.h#L226 in python.

For this usecase, we will have to have immutable metadata constructs in python, no? @nvmkuruc

Reply:
Even when registering validators in python, the keyword arg constructor @roggiezhang-nv provides for validator metadata is probably the most pythonic way to initialize and at least somewhat symmetrical to aggregate initialization of structs with designated initializers in C++.

Reply:
Yup I missed the constructor! Agreed with your statement @nvmkuruc

Reply:
I'll make those properties immutable then.